resolversrv/utils.py

```diff
@@ -15,32 +17,76 @@ def get_records(bibcode, link_type=None, link_sub_type=None):
-                rows = session.query(DataLinks).filter(
-                    and_(DataLinks.bibcode == bibcode, DataLinks.link_type == link_type,
-                         DataLinks.link_sub_type == link_sub_type)).all()
-                msg = "Fetched records for bibcode = %s, link type = %s and link sub type = %s."
-            current_app.logger.debug(msg.format(bibcode, link_type, link_sub_type))
-            results = []
-            for row in rows:
-                results.append(row.toJSON())
-            return results
-    except NoResultFound, e:
+    with current_app.session_scope() as session:
         if link_type is None:
-            msg = "No records found for bibcode = %s."
+            rows = session.query(DataLinks).filter(and_(DataLinks.bibcode == bibcode)).all()
+            current_app.logger.info("Fetched records for bibcode = %s."  % (bibcode))
         elif link_sub_type is None:
-            msg = "No records found for bibcode = %s and link type = %s."
+            rows = session.query(DataLinks).filter(and_(DataLinks.bibcode == bibcode, DataLinks.link_type == link_type)).all()
+            current_app.logger.info("Fetched records for bibcode = %s and link type = %s." % (bibcode, link_type))
         else:
-            msg = "No records found for bibcode = %s, link type = %s and link sub type = %s."
-        current_app.logger.error(msg.format(bibcode, link_type, link_sub_type))
-        current_app.logger.error('Error: ' + e)
-        return None
+            rows = session.query(DataLinks).filter(and_(DataLinks.bibcode == bibcode, DataLinks.link_type == link_type,
+                                                        DataLinks.link_sub_type == link_sub_type)).all()
+            current_app.logger.info("Fetched records for bibcode = %s, link type = %s and link sub type = %s." %
+                                    (bibcode, link_type, link_sub_type))
+
+        if len(rows) == 0:
+            if link_type is None:
+                current_app.logger.error("No records found for bibcode = %s." % (bibcode))
+            elif link_sub_type is None:
+                current_app.logger.error("No records found for bibcode = %s and link type = %s." % (bibcode, link_type))
+            else:
+                current_app.logger.error("No records found for bibcode = %s, link type = %s and link sub type = %s." %
+                                         (bibcode, link_type, link_sub_type))
+            return None
+
+        results = []
+        for row in rows:
+            results.append(row.toJSON())
+        return results
+
+
+def add_records(datalinks_records_list):
+    """
+    upserts records into db
+
+    :param datalinks_records_list:
+    :return: success boolean, plus a status text for retuning error message, if any, to the calling program
+    """
+    if isinstance(datalinks_records_list, DataLinksRecordList):
+        rows = []
+        for i in range(len(datalinks_records_list.datalinks_records)):
```

Comment:
not sure i'm understanding it, why would you examine all possible (2 member) pairs - i.e. permutations?

was it perhpaps meant to be? (combination without repetition)
```
for record in datalinks_record_list.datalinks_records:
  for r in record.data_links_rows:
    row.append(....)
```

Reply:
discussed this in the meeting.